Public/Set-AutomateAPIGeneric.ps1

```diff
@@ -0,0 +1,177 @@
+        [string]
+        $OrderBy,
+
+        [Parameter(Mandatory = $false)]
+        [string]
+        $Condition,
+
+        [Parameter(Mandatory = $false)]
+        [string]
+        $IncludeFields,
+
+        [Parameter(Mandatory = $false)]
+        [string]
+        $ExcludeFields,
+
+        [Parameter(Mandatory = $false)]
+        [string]
+        $IDs,
+
+        [Parameter(Mandatory = $false)]
+        [string]
+        $Expand
+    )
+    
+    begin {
+        #Build the URL to hit
+        $url = ($Script:CWAServer + '/cwa/api/v1/' + $EndPoint)
+
+        #Convert the Body parameters to expected format
+        $BodyString = ($Body | ConvertTo-Json -Compress)
+
+        #Build the Query Up
+        Add-Type -AssemblyName System.Web
+        $Query = [System.Web.HttpUtility]::ParseQueryString([String]::Empty)
+
+        #Put the page size in
+        $Query['pagesize'] = $PageSize
+
+        #Put the condition in
+        if ($Condition) {
+            $Query['condition'] = $condition
+        }
+
+        #Put the orderby in
+        if ($OrderBy) {
+            $Query['orderby'] = $orderby
+        }
+
+        #Include only these fields
+        if ($IncludeFields) {
+            $Query['includefields'] = $IncludeFields
+        }
+
+        #Exclude only these fields
+        if ($ExcludeFields) {
+            $Query['excludefields'] = $ExcludeFields
+        }
+
+        #Include only these IDs
+        if ($IDs) {
+            $Query['ids'] = $IDs
+        }
+
+        #Expands in the returned object
+        if ($Expand) {
+          $Query['expand'] = $Expand
+        }
+        
+        $QueryString = $Query.ToString()
+    }
+    
+    process {
+        if ($AllResults) {
+            $ReturnedResults = @()
+            [System.Collections.ArrayList]$ReturnedResults
+            $i = 0
+            DO {
+                [int]$i += 1
+                $URLNew = "$($url)?page=$($i)&$QueryString"
+                try {
+                    $return = Invoke-RestMethod -Method Post -Uri $URLNew -Headers $script:CWAToken -ContentType "application/json" -Body $BodyString
+                }
+                catch {
+                    Write-Error "Failed to perform Invoke-RestMethod to Automate API with error $_.Exception.Message"
+                }
+
+                $ReturnedResults += ($return)
+            }
+            WHILE ($return.count -gt 0)
+        }
+
+        if ($Page) {
+            $ReturnedResults = @()
+            [System.Collections.ArrayList]$ReturnedResults
+            $URLNew = "$($url)?page=$($Page)&$QueryString"
+            try {
+                $return = Invoke-RestMethod -Method Post -Uri $URLNew -Headers $script:CWAToken -ContentType "application/json" -Body $BodyString
+            }
+            catch {
+                Write-Error "Failed to perform Invoke-RestMethod to Automate API with error $_.Exception.Message"
+            }
+
+            $ReturnedResults += ($return)
+        }
```

Comment:
Can we dry this up?

It seems to me like we could do a do/while loop where we the condition is like while($AllResults -and $ReturnedResults -and $null -eq $Page) in the condition